(2/3) Add Enum for HTLCHandlingFailed Reasons

carlaKC opened the pull request:
This PR adds an enum to represent all possible BOLT 04 error codes, along with some variants that surface additional information that will be useful to the end user.

Depends on #3699
API changes in #3700

carlaKC commented on lightning/src/events/mod.rs:
```diff
@@ -627,6 +627,26 @@ pub enum InboundChannelFunds {
+	/// The HTLC was failed by the local node.
+	Local {
+		/// The BOLT 04 failure code providing a specific failure reason.
+		failure_code: u16
```
Makes sense, will take a look at adding a more detailed enum or adding to `HTLCDestination` 👍

joostjager replied:
>we may not always want to give the sender all the details about a failure (eg if the sender used a private channel but we didn't mean to expose it we might pretend the channel doesnt exist)

Maybe I don't understand this fully, but why would you want to hide information from a local api user?

carlaKC replied:
The private channel example is a case where the BOLT 04 code hides information from the sender of the payment, but we _do_ want to surface that information on the API (comment is from a previous approach that just surfaced the BOLT 04 failure code, so wouldn't provide this additional information).

> why would you want to hide information from a local api user?

I do think there are some cases where the end user _probably_ doesn't care about very detailed errors. For example, `InvalidOnionPayload` - do we really care whether the version is unknown or the onion key is bad when there's nothing that we can do about it?

joostjager replied:
I see, outdated.

About your second remark - the end user is a user of the library, I assume. Maybe hard to know what they care about. But still it's probably more efficient to wait for someone asking for it than it is to add it potentially unnecessarily. In that context interested to know whether the code itself is useful? (posted https://github.com/lightningdevkit/rust-lightning/issues/3541#issuecomment-2728420923)

carlaKC replied:
> About your second remark - the end user is a user of the library, I assume. 

Yeah, whoever is consuming LDK's events 👍 

> But still it's probably more efficient to wait for someone asking for it than it is to add it potentially unnecessarily.

Complaint driven development FTW 😅 

> In that context interested to know whether the failure code itself is useful?

As discussed on call, this does fall into the nice to have category. I do also think it's also a nice readability improvement to get rid of the failure codes scattered through the codebase, reduces the chances of using the wrong values IMO.

carlaKC commented:
I'm going to be out on vacation until 3 March, but will pick this up when I get back!

carlaKC commented:
Pushing* for a conceptual look because I haven't found a way of doing this that I'm _super_ happy with. 
Specifically interested in thoughts on:
- Reason in `HTLCDestination`: already contains information about the failure, so including a `reason` in `HTLCHandlingFailed` would be repetitive for cases like `UnknownNextHop`.
- Surfacing a B04 error code: an alternative approach would be to map all these codes that we don't really care about to another user-friendly enum that summarizes them (eg, `InvalidPayload` covering blinding/hmac/etc)**. Combining the two gets a bit ugly, and I think that it's reasonably usable to have the human readable value for most common errors and fall back to the code for the more obscure/detailed ones.

If this looks okay, I'll go ahead and clean up tests+docs and add a similar for `HTLCDestination::FailedPayment`!

\* Force pushed because this is a different approach; old version that just surfaces B04 failure codes is [here](https://github.com/carlaKC/rust-lightning/commit/4a14debd60647a8ed163aaa43fddc230dde871d6).
** We can't put them in `LocalHTLCFailureReason` because they erase the actual B04 code, so we'd need to track the summary error and the B04 code together (which is doable, but ugly to have API-facing values being carted around internally).

carlaKC commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1258,6 +1258,117 @@ where
+	/// The raw BOLT04 failure code for the HTLC.
+	///
+	/// See: https://github.com/lightning/bolts/blob/master/04-onion-routing.md#returning-errors.
+	FailureCode { code: u16 },
```
> an we just convert every use of error codes to the enum version and then only write it out as a u16 when we go to serialize the failure

Yeah can def do that. I thought that some of the error codes were overly-specific (so not something that we want to expose to the end user), but if that's okay then it's a much simpler approach.

valentinewallace commented:
Played with this a bit locally out of fear of sending you down a dead end... It looks like it's possible to avoid the nested enums and have one big `HTLCFailureReason` enum in the event, which seems like it might be a better API? Could you check out this commit and let me know your thoughts: https://github.com/lightningdevkit/rust-lightning/commit/9f9d532b402c80671f4c21e07c4a76728e87aac4 I didn't fix serialization yet but it seems to otherwise compile.

carlaKC commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1258,6 +1251,165 @@ where
+}
+
+impl LocalHTLCFailureReason {
+	pub(super) fn failure_code(&self) -> u16 {
```
these methods are _quite_ verbose, and it's annoying to have to specify the values for each enum variant twice (`failure_code` and `u16.into()`)

could try to do something [smarter](https://github.com/lightningdevkit/rust-lightning/commit/ecf39f320390c6fee0801785f527776834fa272e) here, but decided to keep it simple - interested in other opinions here!

jkczyz replied:
Could we assign the value when defining the enum using explicit discriminants?

https://doc.rust-lang.org/reference/items/enumerations.html#r-items.enum.unit-only
https://doc.rust-lang.org/reference/items/enumerations.html#r-items.enum.discriminant.explicit.unit-only

joostjager commented on lightning/src/events/mod.rs:
```diff
@@ -465,6 +466,25 @@ impl_writeable_tlv_based_enum_upgradable!(ClosureReason,
+	/// The HTLC was failed locally by our node.
+	Local{
+		/// The reason that our node chose to fail the HTLC.
+		reason: LocalHTLCFailureReason
```
If the goal is to expose all information about a failure to the user, shouldn't the failure parameters be included here too? And possibly the tlv extensions?

Perhaps not in this PR, but just thinking ahead whether it would work.

carlaKC replied:
I think we'd probably put that information in `HTLCHandlingFailed` because it applies across all types of `HTLCDestination` types?

joostjager replied:
I think that's true, except for `HTLCDestinationFailure::Downstream` ?

joostjager replied:
Clarified offline. Can extend the `LocalHTLCFailureReason enum` with the failure type specific parameters and tlv extensions.

carlaKC replied:
Discussed a bit offline: thinking about the case where we want to add extra information to these errors if there's demand for it (eg, adding `htlc_amount` to the `FeeInsufficient` variant).

For `LocalHTLCFailureReason`, we an always add fields to the enum struct. Likewise for `Downstream`, could add a field because we're using `impl_writeable_tlv_based_enum`, but are unlikely to need it as error reasons are encrypted.

carlaKC commented:
Rebased because there were some non-trivial conflicts, and added tests + docs.

In an effort to keep the scope down, decided not to:
- Add a reason to `HTLCDestination::FailedPayment`: to keep LOC change down.
- Refactor `queue_add_htlcs` to surface `LocalHTLCFailureReason` because it gets into the guts of `channel`.
- Replace `msgs::FailMalformedHTLC`'s `failure_code` with `LocalHTLCFailureReason`: change sprawls a bit.

Of all of these, the most compelling to change in this PR is `queue_add_htlcs`, because it'll surface liquidity-related failures which are usually interesting to end users. Happy to pull any of these into this PR, or pick them up in follow ups if there's interest!

valentinewallace commented:
In terms of the API for the event, it looks like it would be best to refactor `HTLCDestination` to just contain a million variants rather than put those million variants nested in an enum within an enum. This makes sense because the current variants are pretty vague/not good, like `FailedPayment`. (Thanks @jkczyz who I consulted on this.) 

This is a pretty significant refactor though so I agree it would be best to split out some preliminary changes into an initial PR. Do you have any instincts for what would be good to start with? Maybe we could start by just returning the big enum from the internal `channel` methods, or split out the PR's current changes that replace the onion failure code with this enum?

valentinewallace commented:
Serialization is also a question here, but I think we can do something similar to what we do for the `legacy_failure_reason` in the `PaymentFailed` event.

wvanlint commented:
Excited for this change! This will help a lot with observability into forwarding failures.

> queue_add_htlc

Definitely interested in exposing the cases inside `queue_add_htlc`, will they otherwise end up as `TemporaryChannelFailure`?

If this change will expose local failures when forwarding HTLCs, is there a plan to expose failures received at the origin node later on as well?

carlaKC commented:
> In terms of the API for the event, it looks like it would be best to refactor HTLCDestination to just contain a million variants rather than put those million variants nested in an enum within an enum. This makes sense because the current variants are pretty vague/not good, like FailedPayment. (Thanks @jkczyz who I consulted on this.)

I think that the advantage of the current `HTLCDestination` types is that they do represent summary of the different "types" of failures. If you're building a dashboard with this API, you can easily get an overview of what's going on just from the 4 variants (vs a large set of `HTLCDestination` variants, where you'd need to classify each yourself to get a summary view) and the nested enum can give you further detail if you want to build more specific dashboards about your own forwards/receives.

Certainly defer to the opinions of the people who've worked on the API longer, but this approach worked reasonably well with [LND](https://lightning.engineering/api-docs/api/lnd/router/subscribe-htlc-events/)/[LNDMon](https://github.com/lightninglabs/lndmon/pull/59). 

> Do you have any instincts for what would be good to start with? 

@wvanlint's suggestion to tackle `queue_add_htlc` seems like a good start to me, since liquidity related failures are usually what people are most interested in.

> Maybe we could start by just returning the big enum from the internal `channel` methods

Just so I'm clear here, this would mean returning `HTLCDestination` from methods like `queue_add_htlc` / `send_htlc` and leaving bolt 04 error code handling as-is (that doesn't seem to belong in `channel.rs`).

> or split out the PR's current changes that replace the onion failure code with this enum?

With the one large `HTLCDestination` enum approach, would we want one large failure code enum as well? 

> This is a pretty significant refactor though so I agree it would be best to split out some preliminary changes into an initial PR. 

This is a much larger refactor than I was expecting, so I also want to make sure that it's going to be worth the review time for the project (esp since I'm new here)? My main goal was to learn more about the codebase, which I have, so it's already been worth it for me!

valentinewallace commented:
Good points there and also appreciate the link to the LND APIs, seeing the example of what works in prod today was helpful! 

Will get back in more detail on your comment soon, but FYI I finally took a look at the updates from the last big push and most of the commits here already look great. So would def be happy to start with that, especially since this PR is pretty big so it's nice to split it up anyway. How does it sound to start with landing most of the commits prior to the last one that actually surfaces anything in the public API?

carlaKC commented:
> Good points there and also appreciate the link to the LND APIs, seeing the example of what works in prod today was helpful!

Opinions on API solidified a bit for me after offline discussion, I think that deprecating `HTLCDestination` like you suggested and splitting up into a "type" and a failure reason could work:

```
enum HTLCHandlingType {
    Forward { node_id: Option<PublicKey>, channel_id: u64 }, 
    Receive { payment_hash: PaymentHash },
    Invalid { requested_forward_scid: u64 },
}

struct HTLCHandlingFailed {
    failure_type: HTLCHandlingType,
    failure_reason: LocalHTLCFailureReason,
}
```

We'd be able to map these two fields to `HTLCDestination` to remain backwards compatible, eg a `HTLCHandlingType::Invalid` + `LocalHTLCFailureReason::UnknownNextPeer` = `HTLCDestiantion::UnknownNextPeer`.

> How does it sound to start with landing most of the commits prior to the last one that actually surfaces anything in the public API?

sgtm! Perhaps also adding a refactor of `queue_add_htlc` if it isn't too many LOC, because that has quite a few interesting failure reasons (will investigate).

valentinewallace commented:
API looks great! Thanks a lot for the patience on this. 

> sgtm! Perhaps also adding a refactor of queue_add_htlc if it isn't too many LOC, because that has quite a few interesting failure reasons (will investigate).

Definitely think that could make sense!

jkczyz reviewed (commented):
Concept ACK

carlaKC commented:
> Could we assign the value when defining the enum using explicit discriminants?

IIUC explicit discriminants require [a unique value per variant](https://doc.rust-lang.org/reference/items/enumerations.html#restrictions), so they won't work here because we're mapping many variants to one error code.

vincenzopalazzo reviewed (commented):
Basical LGTM

valentinewallace commented:
IMO it's a good time to rebase on main this since the first PR landed, this should follow pretty quickly and it looks like the conflicts have accumulated a bit

valentinewallace reviewed (commented):
No major feedback, looking good here!

valentinewallace commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1777,8 +1820,19 @@ impl_writeable_tlv_based_enum!(HTLCFailReasonRepr,
 	(1, Reason) => {
-		(0, failure_code, required),
+		(0, _failure_code, (legacy, u16,
+			|r: &HTLCFailReasonRepr| Some(r.clone()) )),
```
I think this is off because we need to write a `u16` here? Unfortunately our ser macros currently don't enforce that the type being written matches the legacy type :(

valentinewallace commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1777,8 +1820,19 @@ impl_writeable_tlv_based_enum!(HTLCFailReasonRepr,
+		// failure_code was required, and is replaced by reason so any time we do not have a
+		// reason available failure_code will be Some so we can require reason.
+		(4, reason, (default_value,
+			if let Some(code) = _failure_code {
+				let failure_reason: LocalHTLCFailureReason = code.into();
+				RequiredWrapper::from(failure_reason)
+			} else {
+				reason
+			}
```
I think for safety we should return the unknown failure reason here even though (as you note above) the failure_code should always be set in this situation

carlaKC replied:
Changed this to return `InvalidValue` if `reason` is `None` and we don't have a `_failure_code`, similar to the way it's been done for `attribution_data` above.

Open to other approaches here too! Still getting comfortable w/ these macros.

valentinewallace replied:
Makes sense, thanks! Definitely a bit of a learning curve to these macros at times

carlaKC commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1497,7 +1497,19 @@ pub enum LocalHTLCFailureReason {
+	AmountExceedsCapacity,
+	/// The HTLC was failed because zero amount HTLCs are not allowed.
+	ZeroAmount,
+	/// The HTLC was failed because its amount is less than the smallest HTLC that the channel
+	/// can currently accept.
+	LiquidityMinimum,
+	/// The HTLC was failed because its amount is more than then largest HTLC that the channel
+	/// can currently accept.
+	LiquidityMaximum,
```
Originally decided agianst this because I think HTLC(min/max) sounds like we've hit channel policy limits on min/max, when `get_available_balances` actually incorporate all sorts of things (dust, policy, liquidity etc).

Going with the rename + adding more info to the docs instead 👍

carlaKC commented:
Diff for last push [here](https://github.com/lightningdevkit/rust-lightning/compare/f3010bb54c440f56c679ae2d08cc813b2ae9cf87..3a74be62bdb8edfd67651f38322096e7a145eccb), renamed a few enums as requested + fixed up serialization of `HTLCFailReasonRepr`.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4494,68 +4494,56 @@ where
-						reason: failure.data,
-						attribution_data: failure.attribution_data,
-					}));
+	fn construct_pending_htlc_fail_msg<'a>(&self, msg: &msgs::UpdateAddHTLC,
+		counterparty_node_id: &PublicKey, shared_secret: [u8; 32], inbound_err: InboundHTLCErr) -> HTLCFailureMsg {
```
nit: we're trying to adopt rustfmt formatting where possible ahead of actually using it on this file, which would format as such:
```rust
fn construct_pending_htlc_fail_msg<'a>(
    &self, msg: &msgs::UpdateAddHTLC, counterparty_node_id: &PublicKey,
    shared_secret: [u8; 32], inbound_err: InboundHTLCErr,
) -> HTLCFailureMsg {
    ...
}

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5847,16 +5835,14 @@ where
+					Err(inbound_err) => {
 						let htlc_destination = get_failed_htlc_destination(outgoing_scid_opt, update_add_htlc.payment_hash);
-						htlc_fails.push((htlc_fail, htlc_destination));
+						htlc_fails.push((self.construct_pending_htlc_fail_msg(&update_add_htlc, &incoming_counterparty_node_id, shared_secret, inbound_err), htlc_destination));
```
nit: separate this new call into an intermediate var?

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4494,68 +4494,56 @@ where
-		&self, msg: &msgs::UpdateAddHTLC, counterparty_node_id: &PublicKey, shared_secret: [u8; 32],
-		decoded_hop: onion_utils::Hop, allow_underpay: bool,
-		next_packet_pubkey_opt: Option<Result<PublicKey, secp256k1::Error>>,
-	) -> PendingHTLCStatus {
```
Looks like another benefit of this refactor is that we can more easily remove the `PendingHTLCStatus` struct once its legacy usage in `channel.rs` is removed

valentinewallace reviewed (commented):
Just minor tweaks, will start looking for a 2nd reviewer

carlaKC commented on lightning/src/ln/onion_payment.rs:
```diff
@@ -361,7 +361,7 @@ pub(super) fn create_recv_pending_htlc_info(
 		if hashed_preimage != payment_hash {
 			return Err(InboundHTLCErr {
-				err_code: 0x4000|22,
+				reason: LocalHTLCFailureReason::InvalidKeysendPreimage,
```
Pulled out the fix for this in [11a4f3e](https://github.com/lightningdevkit/rust-lightning/pull/3601/commits/11a4f3ef209e0ab9050414a93e124f409f5ade5b) then updated in the refactor 👍

joostjager commented on lightning/src/ln/async_payments_tests.rs:
```diff
@@ -20,7 +20,7 @@ use crate::ln::msgs::{
 };
 use crate::ln::offers_tests;
-use crate::ln::onion_utils::INVALID_ONION_BLINDING;
+use crate::ln::onion_utils::LocalHTLCFailureReason;
```
This commit is just so intense. Getting nervous of reviewing so many trivial changes. I wonder if it is worth to:
* Do the renames from err_code to reason separately in a non-review commit
* Introduce the LocalHTLCFailureReason enum, still unused
* Have a commit that contains just mechanical search/replaces that needs no review. Possibly list all the replaces the were applied.
* Then finish with a commit with the semi-manual leftovers.

Maybe it's too much of an ask though.

carlaKC replied:
I'll take a shot at splitting this up 👍

joostjager replied:
It is incredible how much more reviewable this change makes your PR. Thanks so much! The anxiety is gone 😅

joostjager commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1712,6 +1712,49 @@ impl Into<LocalHTLCFailureReason> for u16 {
 	}
 }
 
+impl_writeable_tlv_based_enum!(LocalHTLCFailureReason,
```
Perhaps some explanation in the commit message or code why we don't just keep writing a u16 code?

joostjager commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1474,6 +1468,250 @@ where
+			| Self::DustLimitHolder
+			| Self::DustLimitCounterparty
+			| Self::FeeSpikeBuffer
+			| Self::ChannelNotReady => UPDATE | 7,
```
This seems to be a fundamental design choice. To allow multiple reasons to map to the same bolt code. Somehow it suggests a layering violation, but perhaps this is just the practical way to do it.

It does make this commit more than the basic 'replace u16 with enum'.

carlaKC replied:
> This seems to be a fundamental design choice. To allow multiple reasons to map to the same bolt code. 

Indeed, this is one variant because it allows us to surface this enum directly on the API. Without that, we'd need some sort of wrapper enum which allows either a direct BOLT 04 mapping or an "enriched" failure that then maps to the BOLT04 mapping. 
```
enum Failure {
    Bolt04(Bolt04Failure)
   Enriched(LocalFailure)
}
```

Since we want to surface additional information about the errors, we have to accept that this many to one mapping has to happen somewhere - it either happens in one enum, or across two. 

The above `Failure` structure also doesn't really make sense to surface to the end user, so we'd then need to map this _further_ to a flat enum, which adds a lot of boiler plate.

> Somehow it suggests a layering violation, but perhaps this is just the practical way to do it.

While it does feel a bit like a layering violation because these values end up on the API, it is also just a more accurate way of representing internal error states. The BOLT04 error codes erase a lot of information that's internally available, so perhaps the actual layering violation is to use wire-level error codes internally? For example, this change also gives us more precise unit testing because we can now assert that the `TemporaryChannelFailure` we're receiving is _actually_ the one we're expecting.

joostjager replied:
>The BOLT04 error codes erase a lot of information that's internally available, so perhaps the actual layering violation is to use wire-level error codes internally?

I love this inversion. Sold.

joostjager commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1546,51 +1784,78 @@ impl_writeable_tlv_based_enum!(HTLCFailReasonRepr,
+			| LocalHTLCFailureReason::DustLimitHolder
+			| LocalHTLCFailureReason::DustLimitCounterparty
+			| LocalHTLCFailureReason::FeeSpikeBuffer
+			| LocalHTLCFailureReason::ChannelNotReady => {
```
Again a bit of that layering friction, at least that is how I perceive it.

joostjager reviewed (commented):
I think - but not sure - that I would have preferred a PR that is just a mechanical substitution of all the u16s for enum values, and inside that a clear separation between code that is only literal search/replace and the rest. Maybe I am overcautious with this change set though.

Then talk about the rest in one or more follow ups.

carlaKC commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1474,6 +1468,250 @@ where
+		} else if self == (BADONION | PERM | 24) {
+			LocalHTLCFailureReason::InvalidOnionBlinding
+		} else {
+			LocalHTLCFailureReason::UnknownFailureCode { code: self }
```
We can always get a failure code that we don't know from peers when we get a `MalformedFail`, so need to be able to handle failure codes we haven't upgraded to understand yet.

carlaKC commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1474,6 +1468,250 @@ where
+
+impl Into<LocalHTLCFailureReason> for u16 {
+	fn into(self) -> LocalHTLCFailureReason {
+		if self == (NODE | 2) {
```
I tried something like this [here](https://github.com/lightningdevkit/rust-lightning/commit/0808e8684800fc2a156a6a02779521743fdc426e). The downside of a map is that you can have a runtime failure to find the code you're looking for if you forget to add something to the map; I felt that the compiler check that `match` has all the variants was more compelling than the readability improvement.

Explicit discriminants were also suggested, but they [require a unique value per variant](https://github.com/lightningdevkit/rust-lightning/pull/3601#issuecomment-2766208229).

carlaKC replied:
Could be: 

```
macro_rules! impl_into_u16_for_local_failure {
    ($($variant:path),+ $(,)?) => {
        impl Into<LocalHTLCFailureReason> for u16 {
            fn into(self) -> LocalHTLCFailureReason {
                $(
                    if self == $variant.failure_code() {
                        return $variant;
                    }
                )+
                LocalHTLCFailureReason::UnknownFailureCode { code: self }
            }
        }
    };
}

impl_into_u16_for_local_failure!(
    LocalHTLCFailureReason::TemporaryNodeFailure,
    LocalHTLCFailureReason::PermanentNodeFailure,
   ... all bolt 04 variants
)
```

jkczyz replied:
Not necessarily advocating this, but you could use the macro to define the enum as well. This would ensure each one is used, whereas otherwise `impl_into_u16_for_local_failure` needs to be updated whenever a new variant is added.

For some creative uses of macros see:
- https://github.com/lightningdevkit/rust-lightning/blob/main/lightning-types/src/features.rs
- https://github.com/lightningdevkit/rust-lightning/blob/main/lightning-custom-message/src/lib.rs
- https://github.com/lightningdevkit/rust-lightning/blob/83e9e80291deccf36432255ba6e49aac46478557/lightning/src/util/ser_macros.rs#L1050

For instance, you could use a pared down version of `composite_custom_message_handler`'s rule to specify each variant as something like:

```rust
macro_rules! failure_reason {
	($($variant:ident($code:expr)),* $(,)*) => {
		// ...
	}
}

failure_reason!(
	TemporaryNodeFailure(UPDATE | 7),
	// ...
)
```

Adding docs may not be as nice. We do something for features like this:

https://github.com/lightningdevkit/rust-lightning/blob/83e9e80291deccf36432255ba6e49aac46478557/lightning-types/src/features.rs#L257

jkczyz commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1474,6 +1468,250 @@ where
+	}
+}
+
+impl Into<LocalHTLCFailureReason> for u16 {
```
Prefer `From` over `Into` as per the `Into` docs: https://doc.rust-lang.org/std/convert/trait.Into.html

carlaKC commented on lightning/src/ln/onion_payment.rs:
```diff
@@ -36,6 +36,14 @@ pub struct InboundHTLCErr {
 }
 
+/// Writes payment data for invalid or unknown payment error code.
+pub (super) fn invalid_payment_err_data(amt_msat: u64, current_height: u32) -> Vec<u8>{
```
Returning the code isn't helpful in the commit that follow this because we start to introduce variants like `InvalidKeysendPreimage` so this function can't always return `IncorrectPaymentDetails`.

carlaKC commented:
Broke the large commit up into smaller ones, no changes in logic except for implementing `From` instead of `Into` as suggested.

joostjager commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1716,8 +1744,15 @@ impl_writeable_tlv_based_enum!(HTLCFailReasonRepr,
 		(2, data, required_vec),
+		// failure_code was required, and is replaced by reason in 0.2 so any time we do not have a
+		// reason available failure_code will be Some and can be expressed as a reason.
+		(4, failure_reason, (default_value, LocalHTLCFailureReason::from(_failure_code.ok_or(DecodeError::InvalidValue)?))),
```
Does this work on downgrades with the even type?

carlaKC replied:
Right, good catch! Even though it's required now, should be an odd tlv to allow downgrade 👍

joostjager commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1861,7 +1896,7 @@ impl HTLCFailReason {
 						hold_times: Vec::new(),
 						#[cfg(any(test, feature = "_test_utils"))]
-						onion_error_code: Some(*failure_code),
+						onion_error_code: Some(failure_reason.failure_code()),
```
Especially for testing, the internal reason could be useful? Not necessarily in this PR.

carlaKC replied:
Good shout, I'll address in a follow up 👍

joostjager commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1650,6 +1650,34 @@ impl From<u16> for LocalHTLCFailureReason {
 }
 
+impl_writeable_tlv_based_enum!(LocalHTLCFailureReason,
+	(1, TemporaryNodeFailure) => {},
```
Commit message clear now. Is this type of empty value tlv structure the way to serialize the enum though? I think I was expecting some internal code for serialization specifically. Might be worth a code comment.

carlaKC replied:
> Is this type of empty value tlv structure the way to serialize the enum though?

afaik yes, it's serializing a struct enum with no fields. If we wanted to add more context to each of these variants, we'd be able to add them as TLVs here. 

There aren't a lot of enums in the codebase that don't have extra data, but `BlindedFailure` is done like this too.

joostjager commented on lightning/src/ln/channel.rs:
```diff
@@ -8854,38 +8843,40 @@ impl<SP: Deref> FundedChannel<SP> where
 		skimmed_fee_msat: Option<u64>, blinding_point: Option<PublicKey>,
 		fee_estimator: &LowerBoundedFeeEstimator<F>, logger: &L
-	) -> Result<Option<msgs::UpdateAddHTLC>, ChannelError>
+	) -> Result<Option<msgs::UpdateAddHTLC>, (LocalHTLCFailureReason, String)>
```
Having detailed reasons now, but still needing a message - isn't that a bit too much now? Could add those additional text fields to the LocalHTLCFailureReason enum perhaps.

carlaKC replied:
Will pick up in a follow up 👍

joostjager reviewed (commented):
Only non-blocking comments. Ready to approve when you consider the PR done.

valentinewallace reviewed (commented):
Echoing Joost, ready to ACK when you are!

carlaKC commented on lightning/src/ln/onion_payment.rs:
```diff
@@ -367,8 +367,8 @@ pub(super) fn create_recv_pending_htlc_info(
-				err_code: 0x4000|22,
-				err_data: Vec::new(),
+				err_code: 0x4000 | 15,
+				err_data: invalid_payment_err_data(amt_msat, current_height),
```
This commit fixes the error code to match what the bilp says.
Adding the data is just a side effect, since it's required for this error code.

joostjager replied:
Ah right, missed that.

carlaKC commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1557,31 +1563,87 @@ pub enum LocalHTLCFailureReason {
 			Self::InvalidOnionKey => BADONION | PERM | 6,
-			Self::TemporaryChannelFailure => UPDATE | 7,
-			Self::PermanentChannelFailure => PERM | 8,
+			Self::TemporaryChannelFailure
```
Thoughts inline, but I think at a high level there's just always going to be some tension having one enum that's used for internal/api/bolt04 which we'll need to live with.

> Could it now be the case that with the specialization to local codes, that some of these values are now only used when received from downstream, but never produced locally (because there's a more specialized variant available)? 

Yes, this is the case for `PermanentNodeFailure`, for example.  But I think that the readability improvement of having this value in tests and future proofing of being able to use a readable value over a u16 is worth it (this PR series fixed 3x cases of incorrect u16 being used, for example).

> Perhaps the generic ones can even be removed in some cases?

If we do this, then BOLT codes received from downstream malformed will all be `UnknownFailureCode` which I think is ugly to surface on the API.

> Wondering if there are other areas that had some kind of if statement that now doesn't cover the new variants while it should. Or is this an unfounded concern?

Meaning, for example (?):
- Code that wants to match on a`TemporaryChannelFailure` 
- It gets `HTLCMaximum` and doesn't understand that translated to `TemporaryChannelFailure`? 

We would have run into that in this PR replacing all the codes (like is done with `is_temporary()`), but certainly something that will need to be handled if we have a future case for matching on a specific BOLT04 variant of the enum.

joostjager replied:
I didn't think of downstream malformed. But indeed, becomes ugly when removing those bolt codes.

>Meaning, for example

Indeed, something like that. Seems also unlikely to happen, because it looks like things are trending towards less specialized handling of failures to prevent routing nodes from gaming the scores. When processing the failure message at the sender, matching happens on pretty broad failure groups.

Perhaps there's a future where the actual code is simply ignored, because it isn't reliable anymore.

But all good then.

valentinewallace commented:
One favor -- we generally try to push the rebase on main separately to make it easier to see the "real" diff since the previous push :) Or post the diff in a comment.

joostjager reviewed (approved):
Final-checked the nits through the rebase, looks good! Thorough, thoughtful work, as always. Thanks!
